tool/tsh/db.go

```diff
@@ -1007,12 +1008,36 @@ func formatDatabaseConfigCommand(clusterFlag string, db tlsca.RouteToDatabase) s
 	return fmt.Sprintf("tsh db config --cluster=%v --format=cmd %v", clusterFlag, db.ServiceName)
 }
 
+// isLocalProxyAlwaysRequired returns true for protocols that always requires
+// an ALPN local proxy.
+func isLocalProxyAlwaysRequired(protocol string) bool {
+	switch protocol {
+	case defaults.ProtocolSQLServer,
+		defaults.ProtocolSnowflake:
+		return true
+	default:
+		return false
+	}
+}
+
+// isLocalProxyRequiredForTLSRouting returns true for protocols that requires
+// an ALPN local proxy only when TLS routing is enabled.
+func isLocalProxyRequiredForTLSRouting(protocol string) bool {
+	// MySQL can use its dedicated port when TLS routing is off, without ALPN
+	// local proxy.
+	return protocol == defaults.ProtocolMySQL
+}
+
 // isLocalProxyRequiredForDatabase returns true if local proxy has to be used
-// for connecting to the provided database. Currently return true if:
-//   - TLS routing is enabled.
-//   - A SQL Server connection always requires a local proxy.
+// for connecting to the provided database.
 func isLocalProxyRequiredForDatabase(tc *client.TeleportClient, db *tlsca.RouteToDatabase) bool {
-	return tc.TLSRoutingEnabled || db.Protocol == defaults.ProtocolSQLServer
+	if isLocalProxyAlwaysRequired(db.Protocol) {
+		return true
+	}
+	if tc.TLSRoutingEnabled && isLocalProxyRequiredForTLSRouting(db.Protocol) {
```

Comment:
This condition doesn't look current. 
Let's say that we want to connect with 
 `tsh db connect` to a  postgres database when Teleport proxy support TLS routing. 
The `isLocalProxyRequiredForTLSRouting`  call will return false so the connection established to a teleport proxy won't leverage TLS routing features like db ping protocol.

Reply:
Good point. How about separating local proxy dependency to two levels `required` and `preferred`?

In `tsh db env/config`, we check if local proxy is required. And in `tsh db connect` we check if local proxy is preferred where local proxy is always used when TLS routing is enabled?

Reply:
>we check if local proxy is preferred where local proxy is always preferred/used when TLS routing is enabled?

Yes, if a TLS routing is enable we should always use a local TLS/ALPN proxy in order to connect to teleport proxy. Where if TLS routing is disable we are leveraging ALPN local proxy to support MsSQL and Snowflake connections (MsSQL client doesn't support x509 certs and SnowSQL cli has some limitation that doesn't allow to set a custom client certs). 


> Or should tsh db connect always use local proxy regardless of multiplex mode?

The local proxy should be alway used for multiplex mode and for MSSQL and Snowflake connection in case if multiplex mode is disabled. 

The `tsh db config/env` commands aims to support GUI access where in case of TLS routing enable we are recomeneing to use `tsh db proxy` or `tsh db proxy --tunnel`  commands instead of `tsh db config/env` 
https://goteleport.com/docs/connect-your-client/gui-clients/#get-connection-information

Reply:
> The tsh db config/env commands aims to support GUI access where in case of TLS routing enable we are recommending to use tsh db proxy or tsh db proxy --tunnel commands instead of tsh db config/env

In the case of the reported issue, the `tsh db config/env` is used for native clients *directly* and in a *script*. And the particular database client used is `pg_dump`. So both `tsh db connect` and `tsh proxy db` is not viable alternative.

Reply:
Reverted `tsh db connect` to always use the local proxy.